### PR TITLE
DEV: Unify mypy options and warn redundant workarounds

### PR DESCRIPTION
**Pull request by exiledkingcc**

### CODE REVIEW

File: .github/workflows/github-ci.yaml [hunk 1]
```diff
@@ -137,6 +126,9 @@ jobs:
       run: |
         echo `ruff --version`
         ruff .
+    - name: Test with mypy
+      run : |
+        mypy pypdf
```
Comment (MartinThoma) [inline .github/workflows/github-ci.yaml:131]: Does mypy behave the same for different Python versions?

Comment (exiledkingcc) [reply .github/workflows/github-ci.yaml:131]: i simply checked py3.8 ~ py3.11, they behaved the same. but for py3.6, probably it's not.

Comment (MartinThoma) [reply .github/workflows/github-ci.yaml:131]: Lets see if I get good answers for this via https://stackoverflow.com/q/77219925/562769

I suspect that it should give the same results, so I tend to agree with this change.

**Comment by MartinThoma** [comment]
Thank you very much for the cleanup :pray:

**Comment by MartinThoma** [comment]
Running mypy for Python 3.10 I now get:

```
pypdf/_protocols.py:10: error: Unused "type: ignore" comment
pypdf/_protocols.py:10: error: Incompatible import of "Protocol" (imported name has type "typing_extensions._SpecialForm", local name has type "typing._SpecialForm")  [misc]
pypdf/_protocols.py:10: note: Error code "misc" not covered by "type: ignore" comment
pypdf/types.py:9: error: Unused "type: ignore" comment
pypdf/types.py:9: error: Incompatible import of "Literal" (imported name has type "typing_extensions._SpecialForm", local name has type "typing._SpecialForm")  [misc]
pypdf/types.py:9: note: Error code "misc" not covered by "type: ignore" comment
pypdf/_xobj_image_helpers.py:22: error: Unused "type: ignore" comment
pypdf/_xobj_image_helpers.py:22: error: Incompatible import of "Literal" (imported name has type "typing_extensions._SpecialForm", local name has type "typing._SpecialForm")  [misc]
pypdf/_xobj_image_helpers.py:22: note: Error code "misc" not covered by "type: ignore" comment
pypdf/_reader.py:323: error: Item "str" of "Union[Union[str, IO[Any]], Path]" has no attribute "mode"  [union-attr]
pypdf/_reader.py:323: error: Item "Path" of "Union[Union[str, IO[Any]], Path]" has no attribute "mode"  [union-attr]
pypdf/_reader.py:[13](https://github.com/py-pdf/pypdf/actions/runs/6439903154/job/17488260473?pr=2242#step:17:14)61: error: Item "IO[Any]" of "Union[BytesIO, IO[Any]]" has no attribute "getbuffer"  [union-attr]
pypdf/_reader.py:[14](https://github.com/py-pdf/pypdf/actions/runs/6439903154/job/17488260473?pr=2242#step:17:15)[15](https://github.com/py-pdf/pypdf/actions/runs/6439903154/job/17488260473?pr=2242#step:17:16): error: Item "IO[Any]" of "Union[BytesIO, IO[Any]]" has no attribute "getbuffer"  [union-attr]
pypdf/_reader.py:[17](https://github.com/py-pdf/pypdf/actions/runs/6439903154/job/17488260473?pr=2242#step:17:18)08: error: "IO[Any]" has no attribute "getbuffer"  [attr-defined]
pypdf/_writer.py:2[20](https://github.com/py-pdf/pypdf/actions/runs/6439903154/job/17488260473?pr=2242#step:17:21)8: error: Argument 1 has incompatible type "Iterable[ObjectDeletionFlag]"; expected "int"  [arg-type]
pypdf/_writer.py:2208: note: Left operand is of type "Union[ObjectDeletionFlag, Iterable[ObjectDeletionFlag]]"
pypdf/_writer.py:2[21](https://github.com/py-pdf/pypdf/actions/runs/6439903154/job/17488260473?pr=2242#step:17:22)0: error: Argument 1 has incompatible type "Iterable[ObjectDeletionFlag]"; expected "int"  [arg-type]
pypdf/_writer.py:[22](https://github.com/py-pdf/pypdf/actions/runs/6439903154/job/17488260473?pr=2242#step:17:23)10: note: Left operand is of type "Union[ObjectDeletionFlag, Iterable[ObjectDeletionFlag]]"
pypdf/_writer.py:2[23](https://github.com/py-pdf/pypdf/actions/runs/6439903154/job/17488260473?pr=2242#step:17:24)7: error: Argument 1 has incompatible type "Iterable[ObjectDeletionFlag]"; expected "int"  [arg-type]
pypdf/_writer.py:2237: note: Left operand is of type "Union[ObjectDeletionFlag, Iterable[ObjectDeletionFlag]]"
Found 14 errors in 5 files (checked 43 source files)
Error: Process completed with exit code 1.
```

I wonder if that has an effect on code using pypdf. If it has, we need to fix it.

**Comment by MartinThoma** [comment]
I'll leave it like that for the moment as your change made the code a lot easier to read. However, if the CI of users fails due to that I'll revert many of the changes.

**Pull request merged**